src/lib/Simplify.hs

```diff
@@ -102,13 +102,15 @@ data WillLinearize = WillLinearize | NoLinearize
 -- This is because top-level functions can have special (AD) rules associated with
 -- them, and we try to preserve their identity for as long as we can. Since the only
 -- elimination form for function is application, we do some extra handling in simplifyApp.
-newtype SimplifyM (i::S) (o::S) (a:: *) = SimplifyM
+newtype SimplifyMP (r::IR) (i::S) (o::S) (a:: *) = SimplifyM
```

Comment:
I don't think we should make the simplification monad polymorphic in the IR parameter. I realize you did this to handle the `catch` traversal, but I think we should do it differently. To do IR-typed simplification properly, we'll want to have three IR variants at play: CoreIR for the IR that goes into simplification, SimpIR for the IR that comes out, and CoreToSimpIR for the substitution.  CoreToSimpIR will allow SimpIR atoms embedded within CoreIR atoms. I've been using this pattern for the Imp traversal and it works well: https://github.com/google-research/dex-lang/pull/1173.